js/src/util/math.ts

```diff
@@ -0,0 +1,139 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+/**
+ * Convert uint16 (logically a float16) to a JS float64.
+ * Adapted from https://gist.github.com/zed/59a413ae2ed4141d2037
+ * @param n {number} the uint16 to convert
+ * @private
+ * @ignore
+ */
+export function uint16ToFloat64(n: number) {
+
+    // magic numbers:
+    // 31 = b011111
+    // 1024 = 2 ** 10
+
+    const sign = n >> 15;
+    const exponent = (n >> 10) & 31;
+    const fraction = n & (1024 - 1);
+
+    if (exponent === 0) {
+        if (fraction === 0) {
+            return sign ? -0 : 0;
+        }
+        return ((-1) ** sign) * (fraction / 1024) * (2 ** -14);
+    }
+    if (exponent === 31) {
+        if (fraction === 0) {
+            return sign ? Number.NEGATIVE_INFINITY : Number.POSITIVE_INFINITY;
+        }
+        return NaN;
+    }
+    return ((-1) ** sign) * (1 + fraction / 1024) * (2 ** (exponent - 15));
+}
+
+/**
+ * Convert a float64 to uint16 (assuming the float64 is logically a float16).
+ * Adapted from https://gist.github.com/zed/59a413ae2ed4141d2037
```

Comment:
Is there a concern with referencing this since it doesn't have a license? It seems like a simple enough thing but I'd hate if we had IP issues because of it

Reply:
It seemed fine because I used it as a reference (not a copy), it's in a different language, the original code is intended to convert binary to float (splitting hairs, I know), and how many different ways are there to implement IEEE 754 half floats?

Reply:
It's definitely not the kind of thing you want to get in the habit of doing. It would be better if the gist author would indicate a license. Technically translating someone's code creates a Derivative Work. Perhaps you can say "algorithm inspired by" instead

Reply:
If you're translating from one language to the other it's problem not losing the original copyright. So if the original code doesn't have a license that's a problem.

Reply:
If you'd like to do it again from a different piece of code, you could start from Numpy, which is under a BSD 3-clause license:
https://github.com/numpy/numpy/blob/master/numpy/core/src/npymath/halffloat.c

Reply:
Thanks for the feedback Wes + Antoine, I'll take a look at it again tonight.